src/Raven.Server/Documents/RemoteAttachmentsSender.cs

```diff
@@ -0,0 +1,425 @@
+                                        var index = Task.WaitAny(uploadTasks);
+                                        try
+                                        {
+                                            var res = await uploadTasks[index];
+                                            HandleUploadTaskResult(res, processed);
+                                        }
+                                        catch (Exception e)
+                                        {
+                                            _exceptions.Add(e);
+                                        }
+                                        finally
+                                        {
+                                            uploadTasks[index] = CreateUploadTaskAsync(directUploaders, currentTime, document);
+                                        }
+
+                                        break;
+                                    default:
+                                        throw new ArgumentOutOfRangeException(nameof(document.Status), $"Document '{document.LowerId} had unexpected status '{document.Status}'.");
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            // Wait for all uploads to complete
+                            try
+                            {
+                                await Task.WhenAll(uploadTasks);
+                            }
+                            catch
+                            {
+                                // we had a failed task, we will handle it below
+                            }
+
+                            foreach (var t in uploadTasks)
+                            {
+                                _token.ThrowIfCancellationRequested();
+
+                                try
+                                {
+                                    var res = await t;
+                                    HandleUploadTaskResult(res, processed);
+                                }
+                                catch (Exception e)
+                                {
+                                    _exceptions.Add(e);
+                                }
+                            }
+                        }
+
+                        if (toRemote.Count != processed.Count)
+                        {
+                            // we had skipped items
+                            if (Logger.IsDebugEnabled)
+                                Logger.Debug($"Skipping upload attachments of '{toRemote.Count - processed.Count:#,#;;0}' documents, Uploaded: {new Size(_totalUploaded, SizeUnit.Bytes)}, read tx open time: '{duration.ElapsedMilliseconds}'. Skipped keys: {string.Join(", ", toRemote.Select(x => processed.Contains(x) == false))}");
+                        }
+
+                        if (processed.Count == 0)
```

Comment:
this isn't added in the upload path, no?

Reply:
added new handling for duplicates see:
https://github.com/ravendb/ravendb/pull/21835